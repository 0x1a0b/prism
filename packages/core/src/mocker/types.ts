import { INode } from '@stoplight/types';

export interface IMockProvider<N = INode, O = IOperationOptions, R = IMockResult<any>> {
  mock: (operation: N, options: O) => Promise<R>;
}

export interface IMockProviderRegistry {
  registerMockProvider(protocol: Protocol.HTTP, provider: IMockProvider): void;
  getMockProvider(protocol: Protocol.HTTP): IMockProvider | undefined;
}

export enum Protocol {
  HTTP = 'http',
  RPC = 'rpc',
}

export interface IOperationOptions {
  protocol: Protocol.HTTP | Protocol.RPC;
  dynamic?: boolean;
}

export interface IMockResult<T = any> {
  data: T;
}
