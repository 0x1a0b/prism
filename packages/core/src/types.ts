/**
 * The below few interfaces (until end comment) should be defined in the @stoplight/types repo
 */

export interface IDisposable {
  dispose: () => void;
}

export type Path = Array<string | number>;

export enum ValidationSeverity {
  ERROR = 'error',
  WARN = 'warn',
  INFO = 'info',
}
export interface IValidation {
  path: Path;
  name: string;
  summary: string;
  severity: ValidationSeverity;
  message: string;
}

// END

export interface IPrism<Input, Output, Config, LoadOpts> {
  load: (opts?: LoadOpts) => Promise<void>;
  process: (input: Input, config?: Config) => Promise<IPrismOutput<Input, Output>>;
}

export interface IPrismConfig {
  mock?: boolean | object;
  security?: boolean | object;
  validate?: boolean | object;
}

export type PrismConfigFactory<C, I> = (input: I) => Promise<C>;
export type PrismConfig<C, I> = C | PrismConfigFactory<C, I>;

export interface ILoader<Options, Resource> {
  load: (opts?: Options) => Promise<Resource[]>;
}

export interface IFilesystemLoaderOpts {
  path: string;
}

export interface IRouter<Resource, Input, Config> {
  route: (opts: { resources: Resource[]; input: Input; config?: Config }) => Promise<Resource>;
}

export interface IForwarder<Resource, Input, Config, Output> {
  forward: (
    opts: { resource?: Resource; input: IPrismInput<Input>; config?: Config }
  ) => Promise<Output>;
}

export interface IMocker<Resource, Input, Config, Output> {
  mock: (
    opts: { resource?: Resource; input: IPrismInput<Input>; config?: Config }
  ) => Promise<Output>;
}

export interface IValidator<Resource, Input, Config, Output> {
  validateInput?: (
    opts: { resource: Resource; input: Input; config?: Config }
  ) => Promise<IValidation[]>;
  validateOutput?: (
    opts: { resource: Resource; output?: Output; config?: Config }
  ) => Promise<IValidation[]>;
}

export interface IPrismComponents<Resource, Input, Output, Config, LoadOpts> {
  config: PrismConfig<Config, Input>;
  loader: ILoader<LoadOpts, Resource>;
  router: IRouter<Resource, Input, Config>;
  forwarder: IForwarder<Resource, Input, Config, Output>;
  mocker: IMocker<Resource, Input, Config, Output>;
  validator: IValidator<Resource, Input, Config, Output>;
}

export interface IPrismInput<I> {
  validations: {
    input: IValidation[];
  };
  data: I;
}

export interface IPrismOutput<I, O> {
  input: I;
  output?: O;
  validations: {
    input: IValidation[];
    output: IValidation[];
  };
}
