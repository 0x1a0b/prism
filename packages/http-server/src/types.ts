import { IFilesystemLoaderOpts } from '@stoplight/prism-core';
import { TPrismHttpComponents, TPrismHttpInstance } from '@stoplight/prism-http';
import { FastifyInstance, ServerOptions } from 'fastify';

export interface IPrismHttpServerOpts {
  loaderOpts?: IFilesystemLoaderOpts;
  fastify?: ServerOptions;
  components?: TPrismHttpComponents;
}

export interface IPrismHttpServer {
  readonly prism: TPrismHttpInstance;
  readonly fastify: FastifyInstance;
  listen: ListenFunc;
}

export type ListenFunc = (port: number, address?: string, backlog?: number) => Promise<string>;
