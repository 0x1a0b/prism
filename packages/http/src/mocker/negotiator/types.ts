import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}
