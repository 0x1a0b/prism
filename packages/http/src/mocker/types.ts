import { IOperationOptions, Protocol } from '@stoplight/prism-core/mocker/types';

export interface IHttpOperationOptions extends IOperationOptions {
  protocol: Protocol.HTTP;
  // Defaults to the lowest 2xx status code
  status: string;
  // Defaults to application/json
  mediaType: string;
  // Defaults to content type
  example?: string;
  // If true mock back response headers
  headers?: boolean;
}
