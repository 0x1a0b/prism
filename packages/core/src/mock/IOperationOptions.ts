export enum Protocol {
  HTTP = 'http',
  RPC = 'rpc',
}

export interface IOperationOptions {
  protocol: Protocol.HTTP | Protocol.RPC;
  dynamic?: boolean;
}
