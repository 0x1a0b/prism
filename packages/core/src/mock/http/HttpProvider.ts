import { IHttpOperation } from '@stoplight/types';
import { IExampleGenerator } from '../generator/IExampleGenerator';
import { IMockProvider } from '../IMockProvider';
import { IMockResult } from '../IMockResult';
import { IHttpOperationConfig, IHttpResponse } from '@stoplight/prism-http/types';

export class HttpProvider
  implements IMockProvider<IHttpOperation, IHttpOperationConfig, IMockResult<IHttpResponse>> {
  constructor(private exampleGenerator: IExampleGenerator<any>) {}

  public async mock(
    resource: IHttpOperation,
    config: IHttpOperationConfig
  ): Promise<IMockResult<IHttpResponse>> {
    const response = resource.responses.find(r => r.code === config.code);
    if (!response) {
      throw new Error(`Response for status code '${config.code}' not found`);
    }

    const content = response.content.find(c => c.mediaType === config.mediaType);
    if (!content) {
      throw new Error(`Content for media type '${config.mediaType}' not found`);
    }

    let body;
    if (config.dynamic) {
      if (!content.schema) {
        throw new Error('Cannot generate response, schema is missing');
      }

      body = await this.exampleGenerator.generate(content.schema, content.mediaType);
    } else {
      const example = content.examples && content.examples.find(e => e.key === config.exampleKey);

      if (!example) {
        throw new Error(`Example for key '${config.exampleKey}' not found`);
      }

      body = example.value;
    }

    return {
      data: {
        statusCode: parseInt(response.code),
        // @todo: HttpHeaderParam[] ?
        headers: { 'Content-type': content.mediaType },
        body,
      },
    };
  }
}
