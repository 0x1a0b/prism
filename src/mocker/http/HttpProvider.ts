import { IHttpOperation } from '@stoplight/types';
import { IMockProvider } from '../IMockProvider';
import { IMockResult } from '../IMockResult';
import { IHttpOperationOptions } from './IHttpOperationOptions';
import { IMockHttpResponse } from './IMockHttpResponse';

export class HttpProvider
  implements IMockProvider<IHttpOperation, IHttpOperationOptions, IMockResult<IMockHttpResponse>> {
  public async mock(
    operation: IHttpOperation,
    options: IHttpOperationOptions
  ): Promise<IMockResult<IMockHttpResponse>> {
    const response = operation.responses.find(r => r.code === options.status);
    if (!response) {
      throw new Error(`Response for status code '${options.status}' not found`);
    }

    const content = response.content.find(c => c.mediaType === options.contentType);
    if (!content) {
      throw new Error(`Content for media type '${options.contentType}' not found`);
    }

    const example = content.examples && content.examples.find(e => e.key === options.example);
    if (!example) {
      throw new Error(`Example for key '${options.example}' not found`);
    }

    return {
      data: {
        status: parseInt(response.code),
        // @todo: HttpHeaderParam[] ?
        headers: { 'Content-type': content.mediaType },
        body: example.value,
      },
    };
  }
}
