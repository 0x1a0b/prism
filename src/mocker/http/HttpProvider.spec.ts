import { IHttpOperation } from '@stoplight/types';
import { HttpProvider } from './HttpProvider';
import { IHttpOperationOptions } from './IHttpOperationOptions';
import { Protocol } from '../IOperationOptions';

describe('HttpProvider', () => {
  let httpProvider: HttpProvider;

  beforeEach(() => {
    httpProvider = new HttpProvider();
  });

  describe('mock()', () => {
    it('mocks', async () => {
      const operation: IHttpOperation = {
        id: 'id',
        method: 'GET',
        path: '/test',
        responses: [
          {
            code: '200',
            content: [
              {
                mediaType: 'application/json',
                examples: [{ key: 'example1', value: '{"test":"value"}' }],
              },
            ],
          },
        ],
      };

      const options: IHttpOperationOptions = {
        protocol: Protocol.HTTP,
        dynamic: false,
        status: '200',
        contentType: 'application/json',
        example: 'example1',
      };

      expect(await httpProvider.mock(operation, options)).toMatchSnapshot();
    });
  });
});
