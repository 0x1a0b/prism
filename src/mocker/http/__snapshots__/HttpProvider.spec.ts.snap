// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpProvider mock() mocks 1`] = `
Object {
  "data": Object {
    "body": "{\\"test\\":\\"value\\"}",
    "headers": Object {
      "Content-type": "application/json",
    },
    "status": 200,
  },
}
`;
